0000-experiment-dataframe.md

```diff
@@ -0,0 +1,320 @@
+the capability to work with pandas data frames in the suggested structure, and perform the following changes:
+- Breakdown the data frame to individual results.
+- Translate the fields
+   - `uuid` <--> `result_id`
+   - `experiment_uuid` <--> `experiment_id`
+   - `type` <--> `name`
+   - `device_name` <--> `backend`
+   - `quality` (db format) <--> `quality` (QE format)
+   - `fit` <--> `value`, `extra`, `source`
+   - TODO: still need to determine the use of `creation_datetime`
+- Efficiently upload/download multiple results from the server. This depends on the
+API supplied by the server; currently mass loading/updating is possible in one API call,
+but mass creation is not supported, meaning the service might have to resort to multithreading.
+
+To preserve backward compatibility, we suggest all data frame handling in `qiskit-ibm-experiment` to be
+indicted using an optional parameter, `dataframe` (bool) which defaults to `False`, inside the methods
+
+- `IBMExperimentService.create_analysis_results()`
+- `IBMExperimentService.update_analysis_results()`
+- `IBMExperimentService.analysis_result()`
+- `IBMExperimentService.analysis_results()`
+
+## Migration plan
+
+Even though this is the change of internal data structure, this may impact community developers who write own experiment on top of our base classes. Thus, the change should be made step by step to guarantee the backward compatibility.
+
+##### qiskit_experiments v.0.6
+Implement the thread safe data class and subclasses for the analysis results. Replace internal analysis results data with the data frame.
+In this version, `ExpeirmentData.analysis_results()` method still returns a list of `AnalysisResult` by default, and calling with the option `dataframe=True` returns the data frame.
+When data frame is disabled, it internally converts data frame into a list of `AnalysisResult` to return.
+
+`ExperimentData.add_analysis_results()` dispatches the add logic.
+The method signature is updated to `results: Optional[Union[AnalysisResult, List[AnalysisResult]]] = None, **kwargs: Any`.
+If results argument is not None, this must be the legacy analysis result data.
+This implies existing analysis classes can still return this legacy object.
+Legacy result objects are internally converted into a single row of the data frame and added.
+If there are only keyword arguments, we can directly populate the analysis result data frame and bypasses generation of redundant `AnalysisResult` which no longer exists in the `ExperimentData`.
+Existing analysis classes in the Qiskit Experiments namespace will switch to this kwargs style, namely, it must directly populate analysis data from `_run_analysis` method.
+
+`ExperimentData.artifacts()` and `.add_artifacts()` methods are newly implemented with new protected member `ExpeirmentData._artifacts` which is a thread safe list of artifacts.
```

Comment:
If we define an artifact class in the service, could we migrate figure handling to use artifacts? Maybe this requires too much change of the web UI to do in one go

Reply:
^ @gadial

Reply:
The service already know how to handle figures, and should definitely be able to handle artifacts as well. I agree that delegating as much possible code to the service is preferable.